Respect dialog modality and fix a regression in wallet unlock

hebasto opened the pull request:
As pointed in bitcoin/bitcoin#23790 a regression in wallet unlock was introduced in bitcoin-core/gui#336 when a synchronous `AskPassphraseDialog` has been replaced with an asynchronous one.

This PR reverts a call back to a synchronous mode.

To make synchronous dialogs behave nice during shutdown some additional changes were made.

Please note that disabling the tray icon menu when a modal dialog is active is useful itself as on master (4ad59042b359f473d5888ecee0c9288dcf98f1c9) it is possible to switch to the "Receive" tab while the GUI is waiting for a password for the "Send" tab:

![Screenshot from 2021-12-17 18-59-51](https://user-images.githubusercontent.com/32963518/146580710-0a755f24-a166-414b-be60-7863232ac778.png)

This is confusing and must be avoided.

Fixes bitcoin/bitcoin#23790.

prayank23 commented:
I was unable to reproduce the issue mentioned in https://github.com/bitcoin/bitcoin/issues/23790 using PR branch, so issue might be fixed. There was a password prompt each time I tried to send bitcoin from an encrypted wallet.

![image](https://user-images.githubusercontent.com/13405205/146591134-bc9aa397-f59c-48ea-8160-2cabff5b2afc.png)

Would like to test few more things and go through each commit since there are 11 commits in this PR before ACK

willyko commented:
testedACK for the issue mentioned in https://github.com/bitcoin/bitcoin/issues/23790

hebasto commented:
Updated 702a243d34f8ef1ca1789ee0c9ee70aef7ec821e -> 80e7e303262e149df78cc8f0eb3ce357b03184ca ([pr509.01](https://github.com/hebasto/gui/commits/pr509.01) -> [pr509.02](https://github.com/hebasto/gui/commits/pr509.02), [diff](https://github.com/hebasto/gui/compare/pr509.01..pr509.02)).

Dropped [`QAction::setStatusTip()`](https://doc.qt.io/qt-5/qaction.html#statusTip-prop) calls:
- status tip works only for top-level parent widget, tray icon's `QMenu` instance is not a top-level one
- even on master this feature does not work consistently across all supported platforms
- the tray icon context menu is not local to the main window status bar, so status tips are likely to be unnoticed by users

prayank23 commented:
Is this error related to this PR in any way? 
![image](https://user-images.githubusercontent.com/13405205/146955197-3ee8ba5e-df08-4396-a06c-ae13a2107c0d.png)

I get this when I try to load a PSBT for signing. This does not happen in v22.0

prayank23 reviewed (approved):
tACK https://github.com/bitcoin-core/gui/pull/509/commits/80e7e303262e149df78cc8f0eb3ce357b03184ca

This PR fixes the bug reported in https://github.com/bitcoin/bitcoin/issues/23790 and I did not find any issues with the code or testing other things.

prayank23 commented:
There is some error in CI though which I missed earlier: https://github.com/bitcoin-core/gui/pull/509/checks?check_run_id=4570631121

hebasto commented:
Updated 80e7e303262e149df78cc8f0eb3ce357b03184ca -> 356e21f68dc953b2efa95fe11954266f655e3eed ([pr509.02](https://github.com/hebasto/gui/commits/pr509.02) -> [pr509.03](https://github.com/hebasto/gui/commits/pr509.03), [diff](https://github.com/hebasto/gui/compare/pr509.02..pr509.03)).

Silenced `-Wunused-variable` warnings for macOS.

prayank23 reviewed (approved):
reACK https://github.com/bitcoin-core/gui/pull/509/commits/356e21f68dc953b2efa95fe11954266f655e3eed

promag commented:
@hebasto have you tried to fix the regression without reverting the other change?

hebasto commented:
@promag 
> @hebasto have you tried to fix the regression without reverting the other change?

Yes. Although, did not finish it. Changes got more and more evolved.

promag commented:
@hebasto I can see why, there are multiple cases. Not sure how I didn't  catch `unlockWallet` blocking case on https://github.com/bitcoin-core/gui/pull/336#pullrequestreview-766699378 review.

I think the best course of action is to revert and follow up with necessary refactors.

Concept ACK.

w0xlt reviewed (approved):
tACK https://github.com/bitcoin-core/gui/pull/509/commits/356e21f68dc953b2efa95fe11954266f655e3eed

Verified that the error mentioned in https://github.com/bitcoin/bitcoin/issues/23790 happens in master branch and this PR fixes it.

Sjors commented on src/qt/bitcoingui.cpp:
```diff
@@ -804,16 +804,26 @@ void BitcoinGUI::createTrayIconMenu()
-                sign_action->setEnabled(signMessageAction->isEnabled());
-                verify_action->setEnabled(verifyMessageAction->isEnabled());
+            if (QApplication::activeModalWidget()) {
+                for (QAction* a : trayIconMenu.get()->actions()) {
```
97692afeb2849b3bd673b2908fd91473b37dbbbb: I don't see any difference in behavior on macOS; before and after this commit, if I open coin selection, it disables everything except the macOS standard items (Options, Show all windows, Hide, Stop)

Sjors commented on src/qt/bitcoin.cpp:
```diff
@@ -265,7 +265,11 @@ void BitcoinApplication::createWindow(const NetworkStyle *networkStyle)
 
     pollShutdownTimer = new QTimer(window);
-    connect(pollShutdownTimer, &QTimer::timeout, window, &BitcoinGUI::detectShutdown);
+    connect(pollShutdownTimer, &QTimer::timeout, [this]{
```
cda881a9ced639d5f22031cf334140a151d9b505 : on macOS when I'm in coin control, before and after this commit, Command + Q is ignored (not delayed). Same with Stop from the tray menu.

But when I hit ctrl + c in the terminal, before this commit it would shut QT down, with this commit it indeed delays the shutdown until the dialog is closed.

Sjors reviewed (commented):
a632303a67e6642aa793af1e74522801b5a12093 indeed fixes the _no lock after send_ regression for me. But it introduces some (much less problematic) regressions, at least on macOS.

Sjors reviewed (approved):
tACK 74e638afb6e310ef55b07b74fcf388a82030dacc

hebasto commented on src/qt/bitcoingui.cpp:
```diff
@@ -790,7 +785,7 @@ void BitcoinGUI::createTrayIconMenu()
 
     // Note: On macOS, the Dock icon's menu already has Show / Hide action.
-    trayIconMenu->addAction(toggleHideAction);
+    QAction* show_hide_action = trayIconMenu->addAction(tr("Show / &Hide"), this, &BitcoinGUI::toggleHidden);
```
Commit message updated.

hebasto commented:
@Sjors 

> you may want to delay this keyboard shortcut, because we're past the translation cut-off?

Change is small, so it's okay to get it merged before string freeze (now we are in "soft translation string freeze").

@prayank23 @w0xlt @promag 

Mind having another look into this PR?

hebasto commented:
On macOS another behavior change noticed. On master, due to https://github.com/bitcoin-core/gui/blob/b7942c94824e4e06489417ee587d1f3ac631fce0/src/qt/bitcoingui.cpp#L316 `optionsAction` has forcibly been hidden in the Dock menu:

![Screenshot from 2022-02-09 06-49-29](https://user-images.githubusercontent.com/32963518/153125150-b274b35e-b72a-4714-8f0a-79659b253008.png)

With this PR `options_action` has no `QAction::PreferencesRole` set, and it is active:

![Screenshot from 2022-02-09 06-56-51](https://user-images.githubusercontent.com/32963518/153125362-840ab29a-920d-4d6a-a3dd-ce2c16c1eab9.png)

@Sjors What do think about this behavior change?

Sjors commented:
It seems confusing to have "Options..." and "Options ->" in a single dropdown. I don't care deeply, but maybe just keep the original behavior?

prayank23 commented:
> @prayank23 @w0xlt @promag
> 
> Mind having another look into this PR?

I don't care about UI issues with Mac but this pull request addresses an issue which affects security on every platform, works as expected so I have no issues with it. Would have been better if reverting changes were simple and less line of code in this PR. Hopefully we don't ship v23.0 with a known issue that affects security.

ACK https://github.com/bitcoin-core/gui/pull/509/commits/74e638afb6e310ef55b07b74fcf388a82030dacc

hebasto commented:
Updated 74e638afb6e310ef55b07b74fcf388a82030dacc -> f730bd7d580502ae3c3b5953ada3724b59f5cd9b ([pr509.04](https://github.com/hebasto/gui/commits/pr509.04) -> [pr509.05](https://github.com/hebasto/gui/commits/pr509.05), [diff](https://github.com/hebasto/gui/compare/pr509.04..pr509.05)).

> It seems confusing to have "Options..." and "Options ->" in a single dropdown. I don't care deeply, but maybe just keep the original behavior?

Fixed.

hebasto commented:
@prayank23 @Sjors 

Would you mind re-acking?

prayank23 reviewed (approved):
tACK https://github.com/bitcoin-core/gui/pull/509/commits/f730bd7d580502ae3c3b5953ada3724b59f5cd9b

Tried on Fedora and found no issues.


![image](https://user-images.githubusercontent.com/13405205/153855880-81450e58-f32d-4b91-b0e6-c952c01fa533.png)

Sjors commented:
Post merge tACK, [only](https://github.com/hebasto/gui/compare/pr509.04..pr509.05) difference is the `PreferencesRole`.

jarolrod commented:
Post merge tACK, I was able to reproduce the original issue and I've done testing to ensure that this PR fixes this.

Sjors commented:
I wonder if this was also a problem when _creating_ a wallet, see #571.
